01_clean.R

```diff
@@ -1,178 +1,136 @@
+wells_prep <- wells_raw %>%
+  mutate(EMS_ID = Well_Num) %>%      
+  group_by(Well_Num1 = Well_Num) %>%
+  nest()
 
-###############################################################################
-# Import and process the groundwater level data. When you download the data from 
-# the GWL tool, name each file with the well's EMS_ID like so: "[EMS_ID].csv",
-# so you can then pass in the EMS_ID to readGWLdata. Save them all in the same 
-# directory (data_raw_dir set below).
-#
-# If you do not want to download the # raw data, a copy of the cleaned, monthly 
-# data is available at: 
-# http://www.data.gov.bc.ca/dbc/catalogue/detail.page?config=dbc&P110=recorduid:179324.
-# Set download_monthly to TRUE if you want to start with the aggregated monthly 
-# data from DataBC
-###############################################################################
+# Create monthly time series for each well
+wells_month <- mutate(wells_prep, data = map(data, ~monthlyValues(.x)))
 
-if (downloadMonthly) {
-  monthlycsv <- "data/GWL_monthly.csv"
-  download.file("https://catalogue.data.gov.bc.ca/dataset/84c06668-8a1e-4629-90a3-051bba903f22/resource/c8c3c750-bf4d-4213-be58-42c6f42510e8/download/gwlmonthly.csv", 
-                destfile = monthlycsv)
-  monthlywells <- read.csv(monthlycsv, stringsAsFactors = FALSE)
-  monthlywells <- split(monthlywells, monthlywells$EMS_ID)
-  
-} else {
-  # Get a list of all the files in the directory
-  wellfiles <- list.files(data_raw_dir, pattern="\\.csv", full.names = TRUE)
-  
-  # import each csv using the 'readGWLdata' function.
-  wells_raw <- lapply(wellfiles, function(x) {
-    tryCatch(
-      readGWLdata(path=x, emsID=gsub("\\.csv", "", basename(x)))
-      , error=function(e) NULL)
-  }
-  )
-  
-  ################################################################################
-  # Create monthly time series for each well.
-  ################################################################################
-  
-  # Use 'monthlyValues' function to create monthly values
-  monthlywells <- lapply(compact(wells_raw), monthlyValues)
-}
-
-# Create full monthly time series, interpolating missing values
-monthlywellts <- lapply(monthlywells, function(x)
-  tryCatch(makeWellTS(x), error=function(e) NULL))
+# Get timeseries interpolate over missing values
+wells_ts <- mutate(wells_month, data = map(data, ~makeWellTS(.x)))
 
 # Remove consecutive strings of missing (interpolated) values from the
 # beginning and end of each time series
-monthlyts_strip <- lapply(compact(monthlywellts), function(x) {
-  start_end <- trimConsRuns(x$nReadings, val=0, head=0.1, tail=0.9, n_consec=4)
-  x[start_end$start:start_end$end,]
-})
 
-# Combine the list of data frames into a single long data frame. This data frame 
-# can then be converted to annual data and analysed.
-monthlywells.ts <- rbind_all(monthlyts_strip)
+wells_ts_trimmed <- mutate(wells_ts, 
```

Comment:
Good point. I think this would work well: it makes sense statistically, it simplifies the code, and it might even speed things up a bit. Also a nice option to include in `bcgroundwater`.